File: python/0078_Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        subsets = []
        self.dfs(nums, 0, [], subsets)
        return subsets
    
    def dfs(self, nums, index, subset, subsets):
        subsets.append(list(subset))
        
        for i in range(index, len(nums)):
            subset.append(nums[i])
            self.dfs(nums, i + 1, subset, subsets)
            subset.pop()

File: cpp/0078_Subsets.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> sub;
        dfs(0, nums, sub, ans);
        return ans;
    }
    void dfs(int i, vector<int>& nums, vector<int>& sub, vector<vector<int>>& ans) {
        ans.push_back(sub);
        for (int j = i; j < nums.size(); j++) {
            sub.push_back(nums[j]);
            dfs(j + 1, nums, sub, ans);
            sub.pop_back();
        }
    }
};